packages/async-storage/webui/src/usePluginStore.ts

```diff
@@ -0,0 +1,150 @@
+import type { KeyValuePair } from '@react-native-async-storage/async-storage/lib/typescript/types';
+import { App } from 'antd';
+import { useDevToolsPluginClient, type EventSubscription } from 'expo/devtools';
+import { useCallback, useEffect, useState } from 'react';
+import { Method, MethodAck } from '../../methods';
+
+interface Deferred<T> {
+  promise: Promise<T>;
+  resolve: (value: T | PromiseLike<T>) => void;
+  reject: (reason?: any) => void;
+}
+
+const methodAck: Record<Method, MethodAck> = {
+  getAll: 'ack:getAll',
+  set: 'ack:set',
+  remove: 'ack:remove',
+};
+
+const promiseMap = new Map<string, Deferred<any>>();
+
+export function usePluginStore(onError: (error: unknown) => void) {
+  const client = useDevToolsPluginClient('async-storage');
+
+  const { message, notification } = App.useApp();
+
+  const [connected, setConnected] = useState(false);
+  useEffect(() => {
+    const interval = setInterval(() => {
+      setConnected(client?.isConnected() ?? false);
+    }, 1000);
+    return () => clearInterval(interval);
+  }, [client]);
```

Comment:
is the loop only wait for the initial connected event?
just curious about it and think if that's anything we can help from the devtools plugin api level.

Reply:
Pretty much yeah, the only expansion I can think of would be something like a `useConnection` hook that reports information about the connection (could be just isConnected, but maybe also error states?). Probably a cleaner way to handle the check with the current API though, I'll give it some thought

Reply:
i will also think for a proper solution for this, trying to re-render the hook when it's connected.

in the meantime, the downside of setInterval here is that even after connected, there is still a 1s loop running.
how about this?
```suggestion
  useEffect(() => {
    let interval = setInterval(() => {
      if (client?.isConnected()) {
        if (interval != null) {
          clearInterval(interval);
          interval = null;
        }
        setConnected(true);
      }
    }, 1000);
    return () => { if (interval != null) { clearInterval(interval); interval = null; } };
  }, [client]);
```

Reply:
I like it, yeah probably should've gone for that from the start tbh

Reply:
i've revisited my api implementation. when `useDevToolsPluginClient()` returns, the websocket will be connected already. so we don't really need the timer to wait for connection.

i try to clone your branch and modify with the following patch. what do you think and is that aligned to your design?
```diff
diff --git a/packages/async-storage/webui/src/usePluginStore.ts b/packages/async-storage/webui/src/usePluginStore.ts
index ab9b502..319a2eb 100644
--- a/packages/async-storage/webui/src/usePluginStore.ts
+++ b/packages/async-storage/webui/src/usePluginStore.ts
@@ -23,24 +23,10 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
   const { message, notification } = App.useApp();
 
-  const [connected, setConnected] = useState(false);
-  useEffect(() => {
-    let interval = setInterval(() => {
-      if (client?.isConnected()) {
-        if (interval != null) {
-          clearInterval(interval);
-          interval = null;
-        }
-        setConnected(true);
-      }
-    }, 1000);
-    return () => { if (interval != null) { clearInterval(interval); interval = null; } };
-  }, [client]);
-
   const [entries, setEntries] = useState<readonly { key: string; value: string | null }[]>([]);
 
   const update = useCallback(async () => {
-    if (!client?.isConnected()) {
+    if (client?.isConnected() === false) {
       message.error('Not connected to host');
       return;
     }
@@ -53,12 +39,12 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
   const set = useCallback(
     async (key: string, value: string) => {
-      if (!client?.isConnected()) {
+      if (client?.isConnected() === false) {
         message.error('Not connected to host');
         return;
       }
       try {
-        return client.sendMessage('set', {
+        return client?.sendMessage('set', {
           key,
           value,
         });
@@ -71,7 +57,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
   const remove = useCallback(
     async (key: string) => {
-      if (!client?.isConnected()) {
+      if (client?.isConnected() === false) {
         message.error('Not connected to host');
         return;
       }
@@ -87,7 +73,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
   );
 
   useEffect(() => {
-    if (!client?.isConnected()) {
+    if (client?.isConnected() === false) {
       notification.error({
         message: 'Critical error:\nNot connected to host - please reload the page',
       });
@@ -98,7 +84,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
     try {
       subscriptions.push(
-        client.addMessageListener(
+        client?.addMessageListener(
           methodAck.getAll,
           ({ result }: { result: readonly KeyValuePair[] }) => {
             setEntries(result.map(([key, value]) => ({ key, value })));
@@ -111,7 +97,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
     try {
       subscriptions.push(
-        client.addMessageListener(methodAck.set, () => {
+        client?.addMessageListener(methodAck.set, () => {
           update();
         })
       );
@@ -121,7 +107,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
     try {
       subscriptions.push(
-        client.addMessageListener(methodAck.remove, () => {
+        client?.addMessageListener(methodAck.remove, () => {
           update();
         })
       );
@@ -130,7 +116,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
     }
 
     subscriptions.push(
-      client.addMessageListener('error', ({ error }: { error: unknown }) => {
+      client?.addMessageListener('error', ({ error }: { error: unknown }) => {
         onError(error);
       })
     );
@@ -151,6 +137,6 @@ export function usePluginStore(onError: (error: unknown) => void) {
     update,
     set,
     remove,
-    ready: connected,
+    ready: client?.isConnected() === true,
   };
 }
```

Reply:
Yeah that looks good! Probably a better DX too. The only other weak point of the API I noticed (while I'm thinking of it) is in error handling/reporting, but realistically this is a dev-mode only package so it's probably fine. Thanks for the support on this!